File: back/src/main/java/csh/back/domain/vote/vote/dto/response/VoteFindWithUpdateCountResponse.java
package csh.back.domain.vote.vote.dto.response;

import csh.back.domain.trip.place.dto.response.TripPlaceFindResponse;
import csh.back.domain.trip.timeline.entity.Timeline;
import csh.back.domain.vote.vote.entity.Vote;
import csh.back.domain.vote.vote.enums.VoteStatus;

import java.util.List;

public record VoteFindWithUpdateCountResponse(
        List<VoteFindResponse> voteResults,
        List<TripPlaceFindResponse> wishPlaceFindResponses,
        Integer updateCount,
        String voteStatus,
        Long confirmedPlaceId
) {
    public static VoteFindWithUpdateCountResponse of(
            List<VoteFindResponse> voteFindResponses,
            List<TripPlaceFindResponse> wishPlaceFindResponses,
            int updateCount,
            Vote vote
    ) {
        VoteStatus voteStatus = vote.getStatus();
        Timeline timeLine = vote.getTimeline() != null ? vote.getTimeline() : null;
        Long confirmedPlaceId = timeLine.getTripWishPlace() != null ? timeLine.getTripWishPlace().getId() : null;
        return new VoteFindWithUpdateCountResponse(voteFindResponses, wishPlaceFindResponses, updateCount, voteStatus.getNickname(), confirmedPlaceId);
    }
}

File: back/src/main/kotlin/csh/back/domain/vote/vote/dto/response/VoteFindWithUpdateCountResponse.kt
package csh.back.domain.vote.vote.dto.response

import csh.back.domain.trip.place.dto.response.TripPlaceFindResponse
import csh.back.domain.vote.vote.entity.Vote

data class VoteFindWithUpdateCountResponse(
    val voteResults: List<VoteFindResponse>,
    val wishPlaceFindResponses: List<TripPlaceFindResponse>,
    val updateCount: Int,
    val voteStatus: String,
    val confirmedPlaceId: Long?,
    val isAnonymous: Boolean,
) {
    companion object {
        @JvmStatic
        fun of(
            voteFindResponses: List<VoteFindResponse>,
            wishPlaceFindResponses: List<TripPlaceFindResponse>,
            updateCount: Int,
            vote: Vote,
        ): VoteFindWithUpdateCountResponse {
            val voteStatus = vote.status
            val timeline = vote.timeline
            val confirmedPlaceId = timeline.tripWishPlace?.id
            return VoteFindWithUpdateCountResponse(
                voteFindResponses,
                wishPlaceFindResponses,
                updateCount,
                voteStatus.nickname,
                confirmedPlaceId,
                vote.isAnonymous,
            )
        }
    }
}
